benchmarks/BaselinesAsyncSocket.cpp

```diff
@@ -0,0 +1,248 @@
+  }
+
+  bool closed_{false};
+};
+
+class ClientConnectCallback : public AsyncSocket::ConnectCallback {
+ public:
+  ClientConnectCallback(EventBase& eventBase, size_t loadSize, size_t msgLength)
+      : eventBase_(eventBase), loadSize_(loadSize), msgLength_(msgLength) {}
+
+  void connect() {
+    eventBase_.runInEventBaseThread([this] {
+      socket_.reset(new AsyncSocket(&eventBase_));
+      SocketAddress clientAaddr("::", PORT);
+      socket_->connect(this, clientAaddr);
+    });
+  }
+
+ private:
+  void connectSuccess() noexcept override {
+    TcpWriter writer;
+    writer.startWriting(*socket_, loadSize_, msgLength_);
+    socket_->close();
+    delete this;
+  }
+
+  void connectErr(const AsyncSocketException& ex) noexcept override {
+    LOG(ERROR) << "connectErr: " << ex.what() << " " << ex.getType();
+    delete this;
+  }
+
+  AsyncSocket::UniquePtr socket_;
+  EventBase& eventBase_;
+  const size_t loadSize_;
+  const size_t msgLength_;
+};
+}
+
+static void BM_Baseline_AsyncSocket_SendReceive(
+    size_t loadSize,
+    size_t msgLength,
+    size_t recvLength) {
+  EventBase serverEventBase;
+  auto serverSocket = AsyncServerSocket::newSocket(&serverEventBase);
+
+  ServerAcceptCallback serverCallback(serverEventBase, loadSize, recvLength);
+
+  SocketAddress addr("::", PORT);
+
+  serverSocket->setReusePortEnabled(true);
+  serverSocket->bind(addr);
+  serverSocket->addAcceptCallback(&serverCallback, &serverEventBase);
+  serverSocket->listen(1);
+  serverSocket->startAccepting();
+
+  ScopedEventBaseThread clientThread;
+  auto* clientCallback = new ClientConnectCallback(
+      *clientThread.getEventBase(), loadSize, msgLength);
+  clientCallback->connect();
+
+  serverEventBase.loopForever();
+}
+
+void BM_Baseline_AsyncSocket_Throughput_100MB_s40B_r1024B(unsigned, unsigned) {
+  constexpr size_t loadSizeB = 100 * 1024 * 1024;
+  constexpr size_t sendSizeB = 40;
+  constexpr size_t receiveSizeB = 1024;
+  BM_Baseline_AsyncSocket_SendReceive(loadSizeB, sendSizeB, receiveSizeB);
+}
+void BM_Baseline_AsyncSocket_Throughput_100MB_s40B_r4096B(unsigned, unsigned) {
+  constexpr size_t loadSizeB = 100 * 1024 * 1024;
+  constexpr size_t sendSizeB = 40;
+  constexpr size_t receiveSizeB = 4096;
+  BM_Baseline_AsyncSocket_SendReceive(loadSizeB, sendSizeB, receiveSizeB);
+}
+void BM_Baseline_AsyncSocket_Throughput_100MB_s80B_r4096B(unsigned, unsigned) {
+  constexpr size_t loadSizeB = 100 * 1024 * 1024;
+  constexpr size_t sendSizeB = 80;
+  constexpr size_t receiveSizeB = 4096;
+  BM_Baseline_AsyncSocket_SendReceive(loadSizeB, sendSizeB, receiveSizeB);
+}
+void BM_Baseline_AsyncSocket_Throughput_100MB_s4096B_r4096B(
+    unsigned,
+    unsigned) {
+  constexpr size_t loadSizeB = 100 * 1024 * 1024;
+  constexpr size_t sendSizeB = 4096;
+  constexpr size_t receiveSizeB = 4096;
+  BM_Baseline_AsyncSocket_SendReceive(loadSizeB, sendSizeB, receiveSizeB);
+}
+
+BENCHMARK_PARAM(BM_Baseline_AsyncSocket_Throughput_100MB_s40B_r1024B, 1);
```

Comment:
yeah, I tried that, but the compiler generates this error:
error: function 'BM_Baseline_AsyncSocket_Latency_1M_msgs_32B' has internal linkage but is not defined [-Werror,-Wundefined-internal]
BENCHMARK(BM_Baseline_AsyncSocket_Latency_1M_msgs_32B);